### DISCUSSION TITLE
Exporting complete paperspace into png or pdf

### DISCUSSION THREAD

**Discussion by littlekittel** [mozman/ezdxf#1031] [Q&A]
Hi all,

I am trying to export a complete paperspace into a png or pdf. Everything you see in the paperspace when opening the dxf.
So far I found documentation about exporting certain regions of the model space but not for entire paperspaces.
Is there anything available out of the box?
I guess an alternative could be to export all viewports and than stitch them back together?

Thank you!

**Comment by mozman** [comment]
The [Frontend](https://ezdxf.mozman.at/docs/addons/drawing.html#frontend) class of the `drawing` add-on can export paperspace layouts.

Tutorial for image export: https://ezdxf.mozman.at/docs/tutorials/image_export.html

**Comment by mozman** [comment]
```Python
import pathlib

import ezdxf
from ezdxf.addons.drawing import Frontend, RenderContext
from ezdxf.addons.drawing.config import (
    Configuration,
    BackgroundPolicy,
    ColorPolicy,
    LineweightPolicy,
)
from ezdxf.addons.drawing import pymupdf, layout


def export(filepath: pathlib.Path, layout_names=("Model",), scale=1):
    doc = ezdxf.readfile(filepath)
    for layout_name in layout_names:
        outname = pathlib.Path(filepath.stem + f"-[{layout_name}]" + ".pdf")
        print(outname)
        if layout_name == "Model":
            dxf_layout = doc.modelspace()
            page = layout.Page(
                0,  # auto-detect
                0,  # auto-detect
                layout.Units.mm,  # 1 drawing unit = 1mm
                layout.Margins.all(0),
                max_width=1189,  # limit page width to 1189mm
                max_height=841,  # limit page height to 841mm
            )
            settings = layout.Settings(scale=scale)
        else:
            try:
                dxf_layout = doc.paperspace(layout_name)
            except KeyError:
                print(f"Layout '{layout_name}' not found")
                continue
            page = layout.Page.from_dxf_layout(dxf_layout)
            settings = layout.Settings(
                fit_page=False,
                scale=dxf_layout.get_plot_unit_scale_factor() * scale,
            )

        backend = pymupdf.PyMuPdfBackend()
        config = Configuration(
            background_policy=BackgroundPolicy.WHITE,
            color_policy=ColorPolicy.MONOCHROME_LIGHT_BG,
            lineweight_policy=LineweightPolicy.RELATIVE,
        )

        Frontend(RenderContext(doc), backend, config=config).draw_layout(dxf_layout)
        pdf_bytes = backend.get_pdf_bytes(page, settings=settings)
        outname.write_bytes(pdf_bytes)


if __name__ == "__main__":
    export(pathlib.Path(r"A_000217.dxf"), ["Model", "Layout1"])
```

This script creates the two PDFs:

`A_000217-[Model].pdf`

![image](https://github.com/mozman/ezdxf/assets/4230966/fe5ea5f0-ea2e-4028-a39a-c14c0c9915e2)

`A_000217-[Layout1].pdf` (arbitrary clipping polygons seem to work now)

![image](https://github.com/mozman/ezdxf/assets/4230966/062e9018-c098-4f54-9cd3-a0181a6e7520)

[A_000217.zip](https://github.com/mozman/ezdxf/files/14318201/A_000217.zip)

**Reply by sunblog3** [reply]
The result of running your script is an error
python 3.11  64bit
PyMuPDF== 1.23.19
ezdxf==1.1.4

```python
Traceback (most recent call last):
  File "E:\demo\A_00217.py", line 57, in <module>
    export(pathlib.Path(r"A_000217.dxf"), ["Model"])
  File "E:\demo\A_00217.py", line 52, in export
    pdf_bytes = backend.get_pdf_bytes(page, settings=settings)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\venv\Lib\site-packages\ezdxf\addons\drawing\pymupdf.py", line 117, in get_pdf_bytes
    backend = self._get_replay(page, settings=settings, render_box=render_box)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\venv\Lib\site-packages\ezdxf\addons\drawing\pymupdf.py", line 100, in _get_replay
    player.replay(backend)
  File "E:\venv\Lib\site-packages\ezdxf\addons\drawing\recorder.py", line 241, in replay
    backend.draw_filled_polygon(data, properties)
  File "E:\venv\Lib\site-packages\ezdxf\addons\drawing\pymupdf.py", line 353, in draw_filled_polygon
    shape.draw_polyline(vertices)
  File "E:\venv\Lib\site-packages\fitz\utils.py", line 3299, in draw_polyline
    self.updateRect(p)
  File "E:\venv\Lib\site-packages\fitz\utils.py", line 3258, in updateRect
    self.rect = fitz.Rect(x, x)
                ^^^^^^^^^^^^^^^
  File "E:\venv\Lib\site-packages\fitz\__init__.py", line 10759, in __init__
    self.x0 = float(x0)
              ^^^^^^^^^
TypeError: float() argument must be a string or a real number, not 'NoneType'
```

**Reply by mozman** [reply]
Fixed incompatibility with newer PyMuPDF versions.

**Reply by littlekittel** [reply]
This works great, thank you!
How is the arbitrary polygon applied on the second layout?

**Reply by mozman** [reply]
It's a VIEWPORT with a polygonal clipping path.

**Comment by AshRam98** [comment]
Hi all, Continuing this thread since it is relevant to my query. In the image below, you see a paperspace of a file. There are certain elements frozen in different viewports. I am trying to plot this paperspace as seen. Is there a way to plot the layout, considering the frozen layers? At the moment, plotting this paperspace shows all the elements in all of the viewports regardless of the frozen layers. 

![unnamed](https://github.com/mozman/ezdxf/assets/77656978/ac40bb49-9c0a-454e-b5d4-2c38b2a23a77)

**Reply by mozman** [reply]
The individual state of frozen layers in viewports is not supported.

UPDATE: The feature is implemented but doesn't work for this example, can you post your (zipped) DXF file?

**Reply by mozman** [reply]
I did a test and it works:

BricsCAD:
![image](https://github.com/mozman/ezdxf/assets/4230966/e24a1fe0-b201-4ada-ab77-d6203212e693)

ezdxf view command:

![image](https://github.com/mozman/ezdxf/assets/4230966/03e0e03c-4c49-4bc3-880c-0775ee8fe619)

Added new example file for testing: https://github.com/mozman/ezdxf/blob/master/examples_dxf/viewport_frozen_layers.dxf
